src/main/java/io/neonbee/internal/processor/EntityProcessor.java

```diff
@@ -90,90 +92,99 @@ public void init(OData odata, ServiceMetadata serviceMetadata) {
+    private Entity parseBody(ODataRequest request, UriResourceEntitySet uriResourceEntitySet, ContentType requestFormat)
             throws DeserializerException {
-        EdmEntityType edmEntityType = uriResourceEntitySet.getEntitySet().getEntityType();
-        Entity entity = odata.createDeserializer(requestFormat).entity(request.getBody(), edmEntityType).getEntity();
-        if (UPDATE.equals(action)) {
-            EdmHelper.addKeyPredicateValues(entity, uriResourceEntitySet, routingContext);
-        }
-        handleEntity(edmEntityType, request, action, entity, entityHandler);
+        EdmEntityType entityType = uriResourceEntitySet.getEntitySet().getEntityType();
+        return odata.createDeserializer(requestFormat).entity(request.getBody(), entityType).getEntity();
     }
 
-    void handleEntity(EdmEntityType edmEntityType, ODataRequest request, DataAction action, Entity entity,
-            Handler<AsyncResult<EntityWrapper>> entityHandler) {
+    @VisibleForTesting
+    Future<EntityWrapper> forwardRequest(ODataRequest request, UriResourceEntitySet uriResourceEntitySet, Entity entity,
+            DataAction action, Promise<Void> processPromise) {
+        EdmEntityType entityType = uriResourceEntitySet.getEntitySet().getEntityType();
         String uriPath = request.getRawRequestUri().replaceFirst(request.getRawBaseUri(), EMPTY);
-        Buffer body = new EntityWrapper(edmEntityType.getFullQualifiedName(), entity).toBuffer(vertx);
+        Buffer body = new EntityWrapper(entityType.getFullQualifiedName(), entity).toBuffer(vertx);
         DataQuery query = new DataQuery(action, uriPath, request.getRawQueryPath(), request.getAllHeaders(), body)
                 .addHeader("X-HTTP-Method", request.getMethod().name());
 
-        requestEntity(vertx, new DataRequest(edmEntityType.getFullQualifiedName(), query),
-                new DataContextImpl(routingContext)).onComplete(entityHandler);
+        return requestEntity(vertx, new DataRequest(entityType.getFullQualifiedName(), query),
+                new DataContextImpl(routingContext)).onFailure(processPromise::fail);
     }
 
     @VisibleForTesting
     @SuppressWarnings("PMD.PrematureDeclaration")
-    Handler<AsyncResult<EntityWrapper>> handleReadEntityResult(UriResourceEntitySet uriResourceEntitySet,
-            EdmEntitySet edmEntitySet, UriInfo uriInfo, ODataResponse response, ContentType responseFormat,
-            HttpStatusCode responseStatus) {
-        return asyncEntity -> handleResult(asyncEntity, processPromise -> {
+    Handler<EntityWrapper> handleReadEntityResult(UriResourceEntitySet uriResourceEntitySet, UriInfo uriInfo,
+            ODataResponse response, ContentType responseFormat, HttpStatusCode responseStatus,
+            Promise<Void> processPromise) {
+        return ew -> {
             // The result entity that will be returned
             Entity foundEntity = null;
```

Comment:
why put this not below where the variable will actually used at line 190 (above the first try catch)?